include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h

```diff
@@ -564,6 +566,75 @@ __pattern_minmax_element(_ExecutionPolicy&& __exec, _Range&& __rng, _Compare __c
     return ::std::make_pair(get<0>(__ret), get<1>(__ret));
 }
 
+template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3, typename _Range4, typename _BinaryPredicate, typename _BinaryOperator>
+oneapi::dpl::__internal::__enable_if_hetero_execution_policy<_ExecutionPolicy,
+                                                             oneapi::dpl::__internal::__difference_t<_Range3>>
+__pattern_reduce_by_segment(_ExecutionPolicy&& __exec, _Range1&& __keys, _Range2&& __values, _Range3&& __out_keys,
+                        _Range4&& __out_values, _BinaryPredicate __binary_pred, _BinaryOperator __binary_op)
+{
+    if (__keys.empty())
+        return 0;
+
+    using namespace oneapi::dpl::experimental::ranges;
+
+    using __diff_type = oneapi::dpl::__internal::__difference_t<_Range1>;
+    using __val_type = oneapi::dpl::__internal::__value_t<_Range2>;
+
+    // Round 1: reduce with extra indices added to avoid long segments
+    auto __n = __keys.size();
+    auto __idx = oneapi::dpl::__par_backend_hetero::__internal::__buffer<_ExecutionPolicy, __diff_type>(__exec, __n)
+                     .get_buffer();
+    auto __tmp_out_keys = oneapi::dpl::__par_backend_hetero::__internal::__buffer<_ExecutionPolicy, __val_type>(__exec, __n)
+                     .get_buffer();
+    auto __tmp_out_values = oneapi::dpl::__par_backend_hetero::__internal::__buffer<_ExecutionPolicy, __diff_type>(__exec, __n)
+                     .get_buffer();
+
+    auto __k1 = oneapi::dpl::__ranges::drop_view_simple<_Range1, __diff_type>(__keys, 1);
+    auto __k2 = __keys;
+
+    auto __view1 = zip_view(views::iota(0, __n), __k1, __k2);
+    auto __view2 = zip_view(views::all_write(__tmp_out_keys), views::all_write(__idx));
+
+    // TODO: replace literal with appropriate SYCL query.
+    auto __res = __pattern_copy_if(::std::forward<_ExecutionPolicy>(__exec), __view1, __view2,
+                                   [__n, __binary_pred](const auto& __a) {
+                                       return ::std::get<0>(__a) == __n ||
+                                              ::std::get<0>(__a) % 2048 == 0 || // segment size
```

Comment:
we have to add more description about copnditions.. and TODO like "the hardcoded split should be more "smart" etc"

Reply:
Done.  The TODO isn't needed since I've just introduced use of the workgroup size to replace the hardcoded segment size.